{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "Requirement already satisfied: pyreadr in /data/phil/miniconda3/envs/aPTR/lib/python3.6/site-packages (0.4.4)\n",
      "Requirement already satisfied: pandas>0.24.0 in /data/phil/miniconda3/envs/aPTR/lib/python3.6/site-packages (from pyreadr) (1.1.5)\n",
      "Requirement already satisfied: pytz>=2017.2 in /data/phil/miniconda3/envs/aPTR/lib/python3.6/site-packages (from pandas>0.24.0->pyreadr) (2022.1)\n",
      "Requirement already satisfied: python-dateutil>=2.7.3 in /home/phil/.local/lib/python3.6/site-packages (from pandas>0.24.0->pyreadr) (2.8.2)\n",
      "Requirement already satisfied: numpy>=1.15.4 in /data/phil/miniconda3/envs/aPTR/lib/python3.6/site-packages (from pandas>0.24.0->pyreadr) (1.17.0)\n",
      "Requirement already satisfied: six>=1.5 in /home/phil/.local/lib/python3.6/site-packages (from python-dateutil>=2.7.3->pandas>0.24.0->pyreadr) (1.16.0)\n",
      "Requirement already satisfied: openpyxl in /data/phil/miniconda3/envs/aPTR/lib/python3.6/site-packages (3.1.0)\n",
      "Requirement already satisfied: et-xmlfile in /data/phil/miniconda3/envs/aPTR/lib/python3.6/site-packages (from openpyxl) (1.1.0)\n"
     ]
    }
   ],
   "source": [
    "# Exploring DIABIMMUNE data\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "!pip install pyreadr\n",
    "!pip install openpyxl\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import os\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subjectID</th>\n",
       "      <th>country</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>cohort</th>\n",
       "      <th>gid_wgs</th>\n",
       "      <th>gid_16s</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E000823</td>\n",
       "      <td>FIN</td>\n",
       "      <td>3000150</td>\n",
       "      <td>56</td>\n",
       "      <td>abx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G74397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E000823</td>\n",
       "      <td>FIN</td>\n",
       "      <td>3102720</td>\n",
       "      <td>259</td>\n",
       "      <td>abx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G69867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E000823</td>\n",
       "      <td>FIN</td>\n",
       "      <td>3114328</td>\n",
       "      <td>760</td>\n",
       "      <td>abx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G74423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E000823</td>\n",
       "      <td>FIN</td>\n",
       "      <td>3106240</td>\n",
       "      <td>511</td>\n",
       "      <td>abx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G74306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E000823</td>\n",
       "      <td>FIN</td>\n",
       "      <td>3000042</td>\n",
       "      <td>151</td>\n",
       "      <td>abx</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G69730</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subjectID country  sampleID  age_at_collection cohort gid_wgs gid_16s\n",
       "0   E000823     FIN   3000150                 56    abx     NaN  G74397\n",
       "1   E000823     FIN   3102720                259    abx     NaN  G69867\n",
       "2   E000823     FIN   3114328                760    abx     NaN  G74423\n",
       "3   E000823     FIN   3106240                511    abx     NaN  G74306\n",
       "4   E000823     FIN   3000042                151    abx     NaN  G69730"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get rdata from internet: \n",
    "# result = pyreadr.read_r('data/diabimmune.rda')\n",
    "\n",
    "diabimmune_metadata = pd.read_table(\"./diabimmune.tsv\")\n",
    "diabimmune_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30.128396536279485\n",
      "1009\n"
     ]
    }
   ],
   "source": [
    "diabimmune_both = diabimmune_metadata.dropna()\n",
    "print(len(diabimmune_both) / len(diabimmune_metadata) * 100) # 30% of the data is available\n",
    "print(len(diabimmune_both))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "257\n",
      "3.9260700389105057\n"
     ]
    }
   ],
   "source": [
    "# How many distinct subjects?\n",
    "n_subjects = len(diabimmune_both['subjectID'].unique())\n",
    "print(n_subjects)\n",
    "print(len(diabimmune_both) / n_subjects) # 4 samples per subject, on average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G74454*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G69809*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G68498*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G36025*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G36451*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G35881*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G35867*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G69741*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G68204*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G68224*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G75872*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G75855*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G73882*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G75694*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G75788*\n",
      "zsh:1: no matches found: /home/phil/DATA/diabimmune/reads/G66253*\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'child' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/IPython/utils/_process_posix.py\u001b[0m in \u001b[0;36msystem\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    156\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 157\u001b[0;31m                 \u001b[0mchild\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpexpect\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspawn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'-c'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcmd\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# Vanilla Pexpect\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    158\u001b[0m             \u001b[0mflush\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/pexpect/pty_spawn.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, command, args, timeout, maxread, searchwindowsize, logfile, cwd, env, ignore_sighup, echo, preexec_fn, encoding, codec_errors, dimensions, use_poll)\u001b[0m\n\u001b[1;32m    204\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 205\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_spawn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpreexec_fn\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdimensions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    206\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muse_poll\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0muse_poll\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/pexpect/pty_spawn.py\u001b[0m in \u001b[0;36m_spawn\u001b[0;34m(self, command, args, preexec_fn, dimensions)\u001b[0m\n\u001b[1;32m    303\u001b[0m         self.ptyproc = self._spawnpty(self.args, env=self.env,\n\u001b[0;32m--> 304\u001b[0;31m                                      cwd=self.cwd, **kwargs)\n\u001b[0m\u001b[1;32m    305\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/pexpect/pty_spawn.py\u001b[0m in \u001b[0;36m_spawnpty\u001b[0;34m(self, args, **kwargs)\u001b[0m\n\u001b[1;32m    314\u001b[0m         \u001b[0;34m'''Spawn a pty and return an instance of PtyProcess.'''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 315\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mptyprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPtyProcess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspawn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    316\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/ptyprocess/ptyprocess.py\u001b[0m in \u001b[0;36mspawn\u001b[0;34m(cls, argv, cwd, env, echo, preexec_fn, dimensions, pass_fds)\u001b[0m\n\u001b[1;32m    314\u001b[0m         \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexec_err_pipe_write\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 315\u001b[0;31m         \u001b[0mexec_err_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexec_err_pipe_read\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m4096\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    316\u001b[0m         \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexec_err_pipe_read\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-59a8855e95e5>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdiabimmune_both\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m     \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\" mv ~/DATA/diabimmune/reads/{diabimmune_both.iloc[i]['gid_16s']}* ~/DATA/diabimmune/16s_shared/reads/\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/ipykernel/zmqshell.py\u001b[0m in \u001b[0;36msystem_piped\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    648\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muser_ns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'_exit_code'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    649\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 650\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muser_ns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'_exit_code'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msystem\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar_expand\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdepth\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    651\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    652\u001b[0m     \u001b[0;31m# Ensure new system_piped implementation is used\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.6/site-packages/IPython/utils/_process_posix.py\u001b[0m in \u001b[0;36msystem\u001b[0;34m(self, cmd)\u001b[0m\n\u001b[1;32m    171\u001b[0m             \u001b[0;31m# (the character is known as ETX for 'End of Text', see\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m             \u001b[0;31m# curses.ascii.ETX).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 173\u001b[0;31m             \u001b[0mchild\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msendline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    174\u001b[0m             \u001b[0;31m# Read and print any more output the program might produce on its\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    175\u001b[0m             \u001b[0;31m# way out.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: local variable 'child' referenced before assignment"
     ]
    }
   ],
   "source": [
    "# At this point, I have downloaded all of the 16S reads. They are in the dir\n",
    "# ~/DATA/diabimmune/reads\n",
    "\n",
    "# I will now move these files to ~/DATA/diabimmune/16s_shared/reads\n",
    "\n",
    "! mkdir -p ~/DATA/diabimmune/16s_shared/reads\n",
    "\n",
    "for i in range(len(diabimmune_both)):\n",
    "    ! mv ~/DATA/diabimmune/reads/{diabimmune_both.iloc[i]['gid_16s']}* ~/DATA/diabimmune/16s_shared/reads/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/1009 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-02-12 16:33:50--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G80124_R1.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 711032697 (678M) [application/x-gzip]\n",
      "Saving to: ‘G80124_R1.fastq.gz’\n",
      "\n",
      "G80124_R1.fastq.gz  100%[===================>] 678.09M  19.2MB/s    in 35s     \n",
      "\n",
      "2023-02-12 16:34:26 (19.2 MB/s) - ‘G80124_R1.fastq.gz’ saved [711032697/711032697]\n",
      "\n",
      "--2023-02-12 16:34:26--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G80124_R2.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 689083198 (657M) [application/x-gzip]\n",
      "Saving to: ‘G80124_R2.fastq.gz’\n",
      "\n",
      "G80124_R2.fastq.gz  100%[===================>] 657.16M  19.2MB/s    in 34s     \n",
      "\n",
      "2023-02-12 16:35:01 (19.1 MB/s) - ‘G80124_R2.fastq.gz’ saved [689083198/689083198]\n",
      "\n",
      "--2023-02-12 16:35:01--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G80124_unpaired.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 404 Not Found\n",
      "2023-02-12 16:35:01 ERROR 404: Not Found.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 1/1009 [01:19<22:13:00, 79.35s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-02-12 16:35:10--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G79049_R1.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 139590801 (133M) [application/x-gzip]\n",
      "Saving to: ‘G79049_R1.fastq.gz’\n",
      "\n",
      "G79049_R1.fastq.gz  100%[===================>] 133.12M  19.2MB/s    in 7.0s    \n",
      "\n",
      "2023-02-12 16:35:17 (19.1 MB/s) - ‘G79049_R1.fastq.gz’ saved [139590801/139590801]\n",
      "\n",
      "--2023-02-12 16:35:17--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G79049_R2.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 132976352 (127M) [application/x-gzip]\n",
      "Saving to: ‘G79049_R2.fastq.gz’\n",
      "\n",
      "G79049_R2.fastq.gz  100%[===================>] 126.82M  19.2MB/s    in 6.6s    \n",
      "\n",
      "2023-02-12 16:35:24 (19.1 MB/s) - ‘G79049_R2.fastq.gz’ saved [132976352/132976352]\n",
      "\n",
      "--2023-02-12 16:35:25--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G79049_unpaired.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 404 Not Found\n",
      "2023-02-12 16:35:25 ERROR 404: Not Found.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 2/1009 [01:36<12:01:32, 42.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-02-12 16:35:27--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G78999_R1.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 585215567 (558M) [application/x-gzip]\n",
      "Saving to: ‘G78999_R1.fastq.gz’\n",
      "\n",
      "G78999_R1.fastq.gz    4%[                    ]  26.05M  18.6MB/s               ^C\n",
      "--2023-02-12 16:35:29--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G78999_R2.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 563767264 (538M) [application/x-gzip]\n",
      "Saving to: ‘G78999_R2.fastq.gz’\n",
      "\n",
      "G78999_R2.fastq.gz  100%[===================>] 537.65M  19.1MB/s    in 28s     \n",
      "\n",
      "2023-02-12 16:35:58 (19.1 MB/s) - ‘G78999_R2.fastq.gz’ saved [563767264/563767264]\n",
      "\n",
      "--2023-02-12 16:35:58--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G78999_unpaired.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 404 Not Found\n",
      "2023-02-12 16:35:58 ERROR 404: Not Found.\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 3/1009 [02:12<11:00:52, 39.42s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2023-02-12 16:36:02--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G45069_R1.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 595936467 (568M) [application/x-gzip]\n",
      "Saving to: ‘G45069_R1.fastq.gz’\n",
      "\n",
      "G45069_R1.fastq.gz   31%[=====>              ] 179.98M  18.0MB/s    eta 21s    ^C\n",
      "--2023-02-12 16:36:13--  https://diabimmune.broadinstitute.org/diabimmune/data/20/G45069_R2.fastq.gz\n",
      "Resolving diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)... 69.173.93.60\n",
      "Connecting to diabimmune.broadinstitute.org (diabimmune.broadinstitute.org)|69.173.93.60|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 584477159 (557M) [application/x-gzip]\n",
      "Saving to: ‘G45069_R2.fastq.gz’\n",
      "\n",
      "G45069_R2.fastq.gz   49%[========>           ] 275.51M  19.2MB/s    eta 15s    "
     ]
    }
   ],
   "source": [
    "# We need to download the corresponding WGS reads\n",
    "\n",
    "! mkdir -p ~/DATA/diabimmune/wgs_shared\n",
    "! mkdir -p ~/DATA/diabimmune/wgs_shared/reads\n",
    "\n",
    "for id in tqdm(diabimmune_both['gid_wgs']):\n",
    "    if not os.path.exists(f\"~/DATA/diabimmune/wgs_shared/reads/{id}_R1.fastq.gz\"):\n",
    "        ! wget https://diabimmune.broadinstitute.org/diabimmune/data/20/{id}_R1.fastq.gz\n",
    "    if not os.path.exists(f\"~/DATA/diabimmune/wgs_shared/reads/{id}_R2.fastq.gz\"):\n",
    "        ! wget https://diabimmune.broadinstitute.org/diabimmune/data/20/{id}_R2.fastq.gz\n",
    "    if not os.path.exists(f\"~/DATA/diabimmune/wgs_shared/reads/{id}_unpaired.fastq.gz\"):\n",
    "        ! wget https://diabimmune.broadinstitute.org/diabimmune/data/20/{id}_unpaired.fastq.gz\n",
    "    ! mv {id}_*.fastq.gz ~/DATA/diabimmune/wgs_shared/reads/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "78\n"
     ]
    }
   ],
   "source": [
    "# Which samples did aPTR return a result for?\n",
    "\n",
    "aptr_out = pd.read_table(\n",
    "    \"/home/phil/aptr/experiments/diabimmune/16s_shared/aptr_8223c61e-4f69-4350-afb5-3b096f484af1/inferred_ptrs.tsv\",\n",
    "    dtype={0: str},\n",
    ")\n",
    "aptr_out = aptr_out.set_index(aptr_out.columns[0])\n",
    "aptr_estimates = aptr_out.loc[:,aptr_out.sum(axis=0) > 0].columns \n",
    "print(len(aptr_estimates))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Move corresponding WGS reads to a special directory - 78 samples\n",
    "\n",
    "# First, make a dictionary to convert sample IDs\n",
    "wgs_to_rna = dict(zip(diabimmune_both['gid_wgs'], diabimmune_both['gid_16s']))\n",
    "rna_to_wgs = dict(zip(diabimmune_both['gid_16s'], diabimmune_both['gid_wgs']))\n",
    "\n",
    "! mkdir -p ~/DATA/diabimmune/wgs_aptr/reads\n",
    "\n",
    "for rna_id in aptr_estimates:\n",
    "    rna_id = rna_id.split(\"_\")[0]\n",
    "    wgs_id = rna_to_wgs[rna_id]\n",
    "    ! cp ~/DATA/diabimmune/wgs_shared/reads/{wgs_id}_R1.fastq.gz ~/DATA/diabimmune/wgs_aptr/reads\n",
    "    ! cp ~/DATA/diabimmune/wgs_shared/reads/{wgs_id}_R2.fastq.gz ~/DATA/diabimmune/wgs_aptr/reads"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run coPTR on matched WGS reads\n",
    "This must be run using the `coPTR` conda environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] [Feb 12, 2023 17:07:11] [coptr.read_mapper] Found 16 files totaling 0.0543 GB.\n",
      "[INFO] [Feb 12, 2023 17:07:11] [coptr.read_mapper] Copying FASTA files to coptr-fna-2023-02-12T22:07:11+00:00.fna with prepended genome ids (filenames).\n",
      "[INFO] [Feb 12, 2023 17:07:12] [coptr.read_mapper] Writing 16 reference genome ids to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs.genomes.\n",
      "[INFO] [Feb 12, 2023 17:07:12] [coptr.read_mapper] bowtie2-build coptr-fna-2023-02-12T22:07:11+00:00.fna /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs --threads 1\n",
      "Settings:\n",
      "  Output files: \"/home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  coptr-fna-2023-02-12T22:07:11+00:00.fna\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 14332445\n",
      "Using parameters --bmax 10749334 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 10749334 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 5.73298e+07 (target: 10749333)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 57329783 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:11\n",
      "Returning block of 57329784 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 15926696\n",
      "fchr[G]: 28664326\n",
      "fchr[T]: 41347056\n",
      "fchr[$]: 57329783\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 23312845 bytes to primary EBWT file: /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs.1.bt2\n",
      "Wrote 14332452 bytes to secondary EBWT file: /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 57329783\n",
      "    bwtLen: 57329784\n",
      "    sz: 14332446\n",
      "    bwtSz: 14332446\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 3583112\n",
      "    offsSz: 14332448\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 298593\n",
      "    numLines: 298593\n",
      "    ebwtTotLen: 19109952\n",
      "    ebwtTotSz: 19109952\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:17\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 14332445\n",
      "Using parameters --bmax 10749334 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 10749334 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:01\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 5.73298e+07 (target: 10749333)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 1\n",
      "  No samples; assembling all-inclusive block\n",
      "  Sorting block of length 57329783 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:11\n",
      "Returning block of 57329784 for bucket 1\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 15926696\n",
      "fchr[G]: 28664326\n",
      "fchr[T]: 41347056\n",
      "fchr[$]: 57329783\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 23312845 bytes to primary EBWT file: /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs.rev.1.bt2\n",
      "Wrote 14332452 bytes to secondary EBWT file: /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs.rev.2.bt2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 57329783\n",
      "    bwtLen: 57329784\n",
      "    sz: 14332446\n",
      "    bwtSz: 14332446\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 3583112\n",
      "    offsSz: 14332448\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 298593\n",
      "    numLines: 298593\n",
      "    ebwtTotLen: 19109952\n",
      "    ebwtTotSz: 19109952\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:17\n",
      "[INFO] [Feb 12, 2023 17:07:46] [coptr.read_mapper] Indexed 16 FASTA files for the reference database.\n",
      "[INFO] [Feb 12, 2023 17:07:46] [coptr.read_mapper] Cleaning up coptr-fna-2023-02-12T22:07:11+00:00.fna.\n"
     ]
    }
   ],
   "source": [
    "# First, make an index for the relevant genomes\n",
    "import pandas as pd\n",
    "\n",
    "aptr_out = pd.read_table(\n",
    "    \"/home/phil/aptr/experiments/diabimmune/16s_shared/aptr_8223c61e-4f69-4350-afb5-3b096f484af1/inferred_ptrs.tsv\",\n",
    "    dtype={0: str},\n",
    ")\n",
    "aptr_genomes = aptr_out.iloc[:,0]\n",
    "\n",
    "! mkdir -p ~/DATA/diabimmune/wgs_aptr/coptr_out/seqs\n",
    "! mkdir -p ~/DATA/diabimmune/wgs_aptr/coptr_out/index\n",
    "\n",
    "for genome_id in aptr_genomes:\n",
    "    ! cp ~/aptr/data/seqs/{genome_id}.fna ~/DATA/diabimmune/wgs_aptr/coptr_out/seqs\n",
    "\n",
    "# Create coPTR index\n",
    "! coptr index ~/DATA/diabimmune/wgs_aptr/coptr_out/seqs ~/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFO] [Feb 12, 2023 17:19:46] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45050.sam\n",
      "[INFO] [Feb 12, 2023 17:19:46] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45050_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45050_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7822277 reads; of these:\n",
      "  7822277 (100.00%) were paired; of these:\n",
      "    7460354 (95.37%) aligned concordantly 0 times\n",
      "    336174 (4.30%) aligned concordantly exactly 1 time\n",
      "    25749 (0.33%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7460354 pairs aligned concordantly 0 times; of these:\n",
      "      171313 (2.30%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7289041 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      14578082 mates make up the pairs; of these:\n",
      "        14479497 (99.32%) aligned 0 times\n",
      "        60852 (0.42%) aligned exactly 1 time\n",
      "        37733 (0.26%) aligned >1 times\n",
      "7.45% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:20:57] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45050.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45050.bam.\n",
      "[INFO] [Feb 12, 2023 17:21:05] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45050.sam.\n",
      "[INFO] [Feb 12, 2023 17:21:05] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45054.sam\n",
      "[INFO] [Feb 12, 2023 17:21:05] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45054_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45054_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7945407 reads; of these:\n",
      "  7945407 (100.00%) were paired; of these:\n",
      "    7381168 (92.90%) aligned concordantly 0 times\n",
      "    542229 (6.82%) aligned concordantly exactly 1 time\n",
      "    22010 (0.28%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7381168 pairs aligned concordantly 0 times; of these:\n",
      "      284823 (3.86%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7096345 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      14192690 mates make up the pairs; of these:\n",
      "        14131364 (99.57%) aligned 0 times\n",
      "        29581 (0.21%) aligned exactly 1 time\n",
      "        31745 (0.22%) aligned >1 times\n",
      "11.07% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:22:24] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45054.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45054.bam.\n",
      "[INFO] [Feb 12, 2023 17:22:34] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45054.sam.\n",
      "[INFO] [Feb 12, 2023 17:22:34] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45056.sam\n",
      "[INFO] [Feb 12, 2023 17:22:34] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45056_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45056_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8166853 reads; of these:\n",
      "  8166853 (100.00%) were paired; of these:\n",
      "    6851464 (83.89%) aligned concordantly 0 times\n",
      "    1272982 (15.59%) aligned concordantly exactly 1 time\n",
      "    42407 (0.52%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6851464 pairs aligned concordantly 0 times; of these:\n",
      "      572627 (8.36%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6278837 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      12557674 mates make up the pairs; of these:\n",
      "        12449175 (99.14%) aligned 0 times\n",
      "        59316 (0.47%) aligned exactly 1 time\n",
      "        49183 (0.39%) aligned >1 times\n",
      "23.78% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:24:20] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45056.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45056.bam.\n",
      "[INFO] [Feb 12, 2023 17:24:42] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45056.sam.\n",
      "[INFO] [Feb 12, 2023 17:24:42] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45067.sam\n",
      "[INFO] [Feb 12, 2023 17:24:42] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45067_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45067_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8627898 reads; of these:\n",
      "  8627898 (100.00%) were paired; of these:\n",
      "    8292504 (96.11%) aligned concordantly 0 times\n",
      "    319339 (3.70%) aligned concordantly exactly 1 time\n",
      "    16055 (0.19%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8292504 pairs aligned concordantly 0 times; of these:\n",
      "      141732 (1.71%) aligned discordantly 1 time\n",
      "    ----\n",
      "    8150772 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      16301544 mates make up the pairs; of these:\n",
      "        16250599 (99.69%) aligned 0 times\n",
      "        26214 (0.16%) aligned exactly 1 time\n",
      "        24731 (0.15%) aligned >1 times\n",
      "5.83% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:25:58] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45067.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45067.bam.\n",
      "[INFO] [Feb 12, 2023 17:26:04] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45067.sam.\n",
      "[INFO] [Feb 12, 2023 17:26:04] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45072.sam\n",
      "[INFO] [Feb 12, 2023 17:26:04] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45072_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45072_2.fastq.gz --no-unal -p 12 -k 10\n",
      "9256966 reads; of these:\n",
      "  9256966 (100.00%) were paired; of these:\n",
      "    7778860 (84.03%) aligned concordantly 0 times\n",
      "    1432388 (15.47%) aligned concordantly exactly 1 time\n",
      "    45718 (0.49%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7778860 pairs aligned concordantly 0 times; of these:\n",
      "      573937 (7.38%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7204923 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      14409846 mates make up the pairs; of these:\n",
      "        14280260 (99.10%) aligned 0 times\n",
      "        80624 (0.56%) aligned exactly 1 time\n",
      "        48962 (0.34%) aligned >1 times\n",
      "22.87% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:28:03] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45072.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45072.bam.\n",
      "[INFO] [Feb 12, 2023 17:28:26] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45072.sam.\n",
      "[INFO] [Feb 12, 2023 17:28:26] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45087.sam\n",
      "[INFO] [Feb 12, 2023 17:28:26] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45087_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45087_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8377611 reads; of these:\n",
      "  8377611 (100.00%) were paired; of these:\n",
      "    6826304 (81.48%) aligned concordantly 0 times\n",
      "    1510570 (18.03%) aligned concordantly exactly 1 time\n",
      "    40737 (0.49%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6826304 pairs aligned concordantly 0 times; of these:\n",
      "      562221 (8.24%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6264083 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      12528166 mates make up the pairs; of these:\n",
      "        12418012 (99.12%) aligned 0 times\n",
      "        69488 (0.55%) aligned exactly 1 time\n",
      "        40666 (0.32%) aligned >1 times\n",
      "25.89% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:30:27] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45087.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45087.bam.\n",
      "[INFO] [Feb 12, 2023 17:30:49] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45087.sam.\n",
      "[INFO] [Feb 12, 2023 17:30:49] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45092.sam\n",
      "[INFO] [Feb 12, 2023 17:30:49] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45092_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45092_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8848058 reads; of these:\n",
      "  8848058 (100.00%) were paired; of these:\n",
      "    8145615 (92.06%) aligned concordantly 0 times\n",
      "    672890 (7.60%) aligned concordantly exactly 1 time\n",
      "    29553 (0.33%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8145615 pairs aligned concordantly 0 times; of these:\n",
      "      289439 (3.55%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7856176 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      15712352 mates make up the pairs; of these:\n",
      "        15632064 (99.49%) aligned 0 times\n",
      "        40857 (0.26%) aligned exactly 1 time\n",
      "        39431 (0.25%) aligned >1 times\n",
      "11.66% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:32:19] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45092.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45092.bam.\n",
      "[INFO] [Feb 12, 2023 17:32:31] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45092.sam.\n",
      "[INFO] [Feb 12, 2023 17:32:31] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45100.sam\n",
      "[INFO] [Feb 12, 2023 17:32:31] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45100_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45100_2.fastq.gz --no-unal -p 12 -k 10\n",
      "9316170 reads; of these:\n",
      "  9316170 (100.00%) were paired; of these:\n",
      "    7962016 (85.46%) aligned concordantly 0 times\n",
      "    1316607 (14.13%) aligned concordantly exactly 1 time\n",
      "    37547 (0.40%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7962016 pairs aligned concordantly 0 times; of these:\n",
      "      839688 (10.55%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7122328 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      14244656 mates make up the pairs; of these:\n",
      "        14113883 (99.08%) aligned 0 times\n",
      "        73425 (0.52%) aligned exactly 1 time\n",
      "        57348 (0.40%) aligned >1 times\n",
      "24.25% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:34:32] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45100.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45100.bam.\n",
      "[INFO] [Feb 12, 2023 17:34:56] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45100.sam.\n",
      "[INFO] [Feb 12, 2023 17:34:56] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45106.sam\n",
      "[INFO] [Feb 12, 2023 17:34:56] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45106_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45106_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8312147 reads; of these:\n",
      "  8312147 (100.00%) were paired; of these:\n",
      "    7692677 (92.55%) aligned concordantly 0 times\n",
      "    595760 (7.17%) aligned concordantly exactly 1 time\n",
      "    23710 (0.29%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7692677 pairs aligned concordantly 0 times; of these:\n",
      "      242514 (3.15%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7450163 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      14900326 mates make up the pairs; of these:\n",
      "        14835100 (99.56%) aligned 0 times\n",
      "        33598 (0.23%) aligned exactly 1 time\n",
      "        31628 (0.21%) aligned >1 times\n",
      "10.76% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:36:18] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45106.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45106.bam.\n",
      "[INFO] [Feb 12, 2023 17:36:28] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45106.sam.\n",
      "[INFO] [Feb 12, 2023 17:36:28] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45272.sam\n",
      "[INFO] [Feb 12, 2023 17:36:28] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45272_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G45272_2.fastq.gz --no-unal -p 12 -k 10\n",
      "11602471 reads; of these:\n",
      "  11602471 (100.00%) were paired; of these:\n",
      "    10511403 (90.60%) aligned concordantly 0 times\n",
      "    1043542 (8.99%) aligned concordantly exactly 1 time\n",
      "    47526 (0.41%) aligned concordantly >1 times\n",
      "    ----\n",
      "    10511403 pairs aligned concordantly 0 times; of these:\n",
      "      551350 (5.25%) aligned discordantly 1 time\n",
      "    ----\n",
      "    9960053 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      19920106 mates make up the pairs; of these:\n",
      "        19718527 (98.99%) aligned 0 times\n",
      "        135930 (0.68%) aligned exactly 1 time\n",
      "        65649 (0.33%) aligned >1 times\n",
      "15.02% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:38:38] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45272.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45272.bam.\n",
      "[INFO] [Feb 12, 2023 17:38:57] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45272.sam.\n",
      "[INFO] [Feb 12, 2023 17:38:58] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69166.sam\n",
      "[INFO] [Feb 12, 2023 17:38:58] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69166_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69166_2.fastq.gz --no-unal -p 12 -k 10\n",
      "10180750 reads; of these:\n",
      "  10180750 (100.00%) were paired; of these:\n",
      "    9000612 (88.41%) aligned concordantly 0 times\n",
      "    1147648 (11.27%) aligned concordantly exactly 1 time\n",
      "    32490 (0.32%) aligned concordantly >1 times\n",
      "    ----\n",
      "    9000612 pairs aligned concordantly 0 times; of these:\n",
      "      973533 (10.82%) aligned discordantly 1 time\n",
      "    ----\n",
      "    8027079 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      16054158 mates make up the pairs; of these:\n",
      "        15910087 (99.10%) aligned 0 times\n",
      "        70071 (0.44%) aligned exactly 1 time\n",
      "        74000 (0.46%) aligned >1 times\n",
      "21.86% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:40:59] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69166.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69166.bam.\n",
      "[INFO] [Feb 12, 2023 17:41:22] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69166.sam.\n",
      "[INFO] [Feb 12, 2023 17:41:22] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69167.sam\n",
      "[INFO] [Feb 12, 2023 17:41:22] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69167_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69167_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3878788 reads; of these:\n",
      "  3878788 (100.00%) were paired; of these:\n",
      "    3308725 (85.30%) aligned concordantly 0 times\n",
      "    554244 (14.29%) aligned concordantly exactly 1 time\n",
      "    15819 (0.41%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3308725 pairs aligned concordantly 0 times; of these:\n",
      "      376530 (11.38%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2932195 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5864390 mates make up the pairs; of these:\n",
      "        5805336 (98.99%) aligned 0 times\n",
      "        30976 (0.53%) aligned exactly 1 time\n",
      "        28078 (0.48%) aligned >1 times\n",
      "25.17% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:42:10] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69167.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69167.bam.\n",
      "[INFO] [Feb 12, 2023 17:42:20] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69167.sam.\n",
      "[INFO] [Feb 12, 2023 17:42:20] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69178.sam\n",
      "[INFO] [Feb 12, 2023 17:42:20] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69178_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69178_2.fastq.gz --no-unal -p 12 -k 10\n",
      "5811733 reads; of these:\n",
      "  5811733 (100.00%) were paired; of these:\n",
      "    4627923 (79.63%) aligned concordantly 0 times\n",
      "    1155267 (19.88%) aligned concordantly exactly 1 time\n",
      "    28543 (0.49%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4627923 pairs aligned concordantly 0 times; of these:\n",
      "      574136 (12.41%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4053787 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8107574 mates make up the pairs; of these:\n",
      "        8019281 (98.91%) aligned 0 times\n",
      "        49445 (0.61%) aligned exactly 1 time\n",
      "        38848 (0.48%) aligned >1 times\n",
      "31.01% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:43:42] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69178.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69178.bam.\n",
      "[INFO] [Feb 12, 2023 17:44:00] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69178.sam.\n",
      "[INFO] [Feb 12, 2023 17:44:01] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69187.sam\n",
      "[INFO] [Feb 12, 2023 17:44:01] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69187_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69187_2.fastq.gz --no-unal -p 12 -k 10\n",
      "5345938 reads; of these:\n",
      "  5345938 (100.00%) were paired; of these:\n",
      "    4800198 (89.79%) aligned concordantly 0 times\n",
      "    534875 (10.01%) aligned concordantly exactly 1 time\n",
      "    10865 (0.20%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4800198 pairs aligned concordantly 0 times; of these:\n",
      "      1705466 (35.53%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3094732 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6189464 mates make up the pairs; of these:\n",
      "        6074398 (98.14%) aligned 0 times\n",
      "        38766 (0.63%) aligned exactly 1 time\n",
      "        76300 (1.23%) aligned >1 times\n",
      "43.19% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:45:21] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69187.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69187.bam.\n",
      "[INFO] [Feb 12, 2023 17:45:44] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69187.sam.\n",
      "[INFO] [Feb 12, 2023 17:45:44] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69192.sam\n",
      "[INFO] [Feb 12, 2023 17:45:44] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69192_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69192_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6218700 reads; of these:\n",
      "  6218700 (100.00%) were paired; of these:\n",
      "    4539293 (72.99%) aligned concordantly 0 times\n",
      "    1644583 (26.45%) aligned concordantly exactly 1 time\n",
      "    34824 (0.56%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4539293 pairs aligned concordantly 0 times; of these:\n",
      "      1006595 (22.18%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3532698 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      7065396 mates make up the pairs; of these:\n",
      "        6947779 (98.34%) aligned 0 times\n",
      "        63289 (0.90%) aligned exactly 1 time\n",
      "        54328 (0.77%) aligned >1 times\n",
      "44.14% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:47:29] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69192.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69192.bam.\n",
      "[INFO] [Feb 12, 2023 17:47:55] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69192.sam.\n",
      "[INFO] [Feb 12, 2023 17:47:56] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69193.sam\n",
      "[INFO] [Feb 12, 2023 17:47:56] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69193_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69193_2.fastq.gz --no-unal -p 12 -k 10\n",
      "9495589 reads; of these:\n",
      "  9495589 (100.00%) were paired; of these:\n",
      "    7592814 (79.96%) aligned concordantly 0 times\n",
      "    1859817 (19.59%) aligned concordantly exactly 1 time\n",
      "    42958 (0.45%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7592814 pairs aligned concordantly 0 times; of these:\n",
      "      1238407 (16.31%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6354407 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      12708814 mates make up the pairs; of these:\n",
      "        12565920 (98.88%) aligned 0 times\n",
      "        67264 (0.53%) aligned exactly 1 time\n",
      "        75630 (0.60%) aligned >1 times\n",
      "33.83% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:50:16] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69193.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69193.bam.\n",
      "[INFO] [Feb 12, 2023 17:50:50] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69193.sam.\n",
      "[INFO] [Feb 12, 2023 17:50:51] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69206.sam\n",
      "[INFO] [Feb 12, 2023 17:50:51] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69206_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69206_2.fastq.gz --no-unal -p 12 -k 10\n",
      "12270709 reads; of these:\n",
      "  12270709 (100.00%) were paired; of these:\n",
      "    10723332 (87.39%) aligned concordantly 0 times\n",
      "    1507043 (12.28%) aligned concordantly exactly 1 time\n",
      "    40334 (0.33%) aligned concordantly >1 times\n",
      "    ----\n",
      "    10723332 pairs aligned concordantly 0 times; of these:\n",
      "      1008820 (9.41%) aligned discordantly 1 time\n",
      "    ----\n",
      "    9714512 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      19429024 mates make up the pairs; of these:\n",
      "        19285549 (99.26%) aligned 0 times\n",
      "        71540 (0.37%) aligned exactly 1 time\n",
      "        71935 (0.37%) aligned >1 times\n",
      "21.42% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:53:20] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69206.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69206.bam.\n",
      "[INFO] [Feb 12, 2023 17:53:47] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69206.sam.\n",
      "[INFO] [Feb 12, 2023 17:53:48] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69211.sam\n",
      "[INFO] [Feb 12, 2023 17:53:48] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69211_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69211_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6435454 reads; of these:\n",
      "  6435454 (100.00%) were paired; of these:\n",
      "    5094555 (79.16%) aligned concordantly 0 times\n",
      "    1314007 (20.42%) aligned concordantly exactly 1 time\n",
      "    26892 (0.42%) aligned concordantly >1 times\n",
      "    ----\n",
      "    5094555 pairs aligned concordantly 0 times; of these:\n",
      "      758759 (14.89%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4335796 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8671592 mates make up the pairs; of these:\n",
      "        8603288 (99.21%) aligned 0 times\n",
      "        27532 (0.32%) aligned exactly 1 time\n",
      "        40772 (0.47%) aligned >1 times\n",
      "33.16% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:55:20] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69211.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69211.bam.\n",
      "[INFO] [Feb 12, 2023 17:55:40] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69211.sam.\n",
      "[INFO] [Feb 12, 2023 17:55:40] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69212.sam\n",
      "[INFO] [Feb 12, 2023 17:55:40] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69212_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69212_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3377170 reads; of these:\n",
      "  3377170 (100.00%) were paired; of these:\n",
      "    2646986 (78.38%) aligned concordantly 0 times\n",
      "    714994 (21.17%) aligned concordantly exactly 1 time\n",
      "    15190 (0.45%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2646986 pairs aligned concordantly 0 times; of these:\n",
      "      690559 (26.09%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1956427 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3912854 mates make up the pairs; of these:\n",
      "        3861421 (98.69%) aligned 0 times\n",
      "        16363 (0.42%) aligned exactly 1 time\n",
      "        35070 (0.90%) aligned >1 times\n",
      "42.83% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 17:56:35] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69212.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69212.bam.\n",
      "[INFO] [Feb 12, 2023 17:56:49] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69212.sam.\n",
      "[INFO] [Feb 12, 2023 17:56:49] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69227.sam\n",
      "[INFO] [Feb 12, 2023 17:56:49] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69227_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G69227_2.fastq.gz --no-unal -p 12 -k 10\n",
      "12500195 reads; of these:\n",
      "  12500195 (100.00%) were paired; of these:\n",
      "    8914885 (71.32%) aligned concordantly 0 times\n",
      "    3506032 (28.05%) aligned concordantly exactly 1 time\n",
      "    79278 (0.63%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8914885 pairs aligned concordantly 0 times; of these:\n",
      "      1226506 (13.76%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7688379 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      15376758 mates make up the pairs; of these:\n",
      "        15136064 (98.43%) aligned 0 times\n",
      "        168536 (1.10%) aligned exactly 1 time\n",
      "        72158 (0.47%) aligned >1 times\n",
      "39.46% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:00:18] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69227.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69227.bam.\n",
      "[INFO] [Feb 12, 2023 18:01:08] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69227.sam.\n",
      "[INFO] [Feb 12, 2023 18:01:08] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78487.sam\n",
      "[INFO] [Feb 12, 2023 18:01:08] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78487_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78487_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3211100 reads; of these:\n",
      "  3211100 (100.00%) were paired; of these:\n",
      "    2788473 (86.84%) aligned concordantly 0 times\n",
      "    411901 (12.83%) aligned concordantly exactly 1 time\n",
      "    10726 (0.33%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2788473 pairs aligned concordantly 0 times; of these:\n",
      "      115789 (4.15%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2672684 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5345368 mates make up the pairs; of these:\n",
      "        5312077 (99.38%) aligned 0 times\n",
      "        22839 (0.43%) aligned exactly 1 time\n",
      "        10452 (0.20%) aligned >1 times\n",
      "17.29% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:01:43] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78487.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78487.bam.\n",
      "[INFO] [Feb 12, 2023 18:01:50] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78487.sam.\n",
      "[INFO] [Feb 12, 2023 18:01:50] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78494.sam\n",
      "[INFO] [Feb 12, 2023 18:01:50] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78494_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78494_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2609279 reads; of these:\n",
      "  2609279 (100.00%) were paired; of these:\n",
      "    2380549 (91.23%) aligned concordantly 0 times\n",
      "    221945 (8.51%) aligned concordantly exactly 1 time\n",
      "    6785 (0.26%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2380549 pairs aligned concordantly 0 times; of these:\n",
      "      75835 (3.19%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2304714 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4609428 mates make up the pairs; of these:\n",
      "        4583099 (99.43%) aligned 0 times\n",
      "        18691 (0.41%) aligned exactly 1 time\n",
      "        7638 (0.17%) aligned >1 times\n",
      "12.18% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:02:15] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78494.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78494.bam.\n",
      "[INFO] [Feb 12, 2023 18:02:19] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78494.sam.\n",
      "[INFO] [Feb 12, 2023 18:02:20] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78509.sam\n",
      "[INFO] [Feb 12, 2023 18:02:20] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78509_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78509_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7717047 reads; of these:\n",
      "  7717047 (100.00%) were paired; of these:\n",
      "    6877207 (89.12%) aligned concordantly 0 times\n",
      "    817212 (10.59%) aligned concordantly exactly 1 time\n",
      "    22628 (0.29%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6877207 pairs aligned concordantly 0 times; of these:\n",
      "      295104 (4.29%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6582103 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      13164206 mates make up the pairs; of these:\n",
      "        13096737 (99.49%) aligned 0 times\n",
      "        39779 (0.30%) aligned exactly 1 time\n",
      "        27690 (0.21%) aligned >1 times\n",
      "15.14% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:03:44] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78509.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78509.bam.\n",
      "[INFO] [Feb 12, 2023 18:03:58] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78509.sam.\n",
      "[INFO] [Feb 12, 2023 18:03:58] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78517.sam\n",
      "[INFO] [Feb 12, 2023 18:03:58] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78517_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78517_2.fastq.gz --no-unal -p 12 -k 10\n",
      "4289906 reads; of these:\n",
      "  4289906 (100.00%) were paired; of these:\n",
      "    3146942 (73.36%) aligned concordantly 0 times\n",
      "    1120385 (26.12%) aligned concordantly exactly 1 time\n",
      "    22579 (0.53%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3146942 pairs aligned concordantly 0 times; of these:\n",
      "      292617 (9.30%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2854325 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5708650 mates make up the pairs; of these:\n",
      "        5648834 (98.95%) aligned 0 times\n",
      "        42315 (0.74%) aligned exactly 1 time\n",
      "        17501 (0.31%) aligned >1 times\n",
      "34.16% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:05:03] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78517.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78517.bam.\n",
      "[INFO] [Feb 12, 2023 18:05:18] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78517.sam.\n",
      "[INFO] [Feb 12, 2023 18:05:18] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78528.sam\n",
      "[INFO] [Feb 12, 2023 18:05:18] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78528_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78528_2.fastq.gz --no-unal -p 12 -k 10\n",
      "4756615 reads; of these:\n",
      "  4756615 (100.00%) were paired; of these:\n",
      "    3756776 (78.98%) aligned concordantly 0 times\n",
      "    977873 (20.56%) aligned concordantly exactly 1 time\n",
      "    21966 (0.46%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3756776 pairs aligned concordantly 0 times; of these:\n",
      "      178699 (4.76%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3578077 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      7156154 mates make up the pairs; of these:\n",
      "        7096325 (99.16%) aligned 0 times\n",
      "        45376 (0.63%) aligned exactly 1 time\n",
      "        14453 (0.20%) aligned >1 times\n",
      "25.41% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:06:21] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78528.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78528.bam.\n",
      "[INFO] [Feb 12, 2023 18:06:33] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78528.sam.\n",
      "[INFO] [Feb 12, 2023 18:06:33] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78541.sam\n",
      "[INFO] [Feb 12, 2023 18:06:33] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78541_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78541_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6296118 reads; of these:\n",
      "  6296118 (100.00%) were paired; of these:\n",
      "    5070023 (80.53%) aligned concordantly 0 times\n",
      "    1198197 (19.03%) aligned concordantly exactly 1 time\n",
      "    27898 (0.44%) aligned concordantly >1 times\n",
      "    ----\n",
      "    5070023 pairs aligned concordantly 0 times; of these:\n",
      "      801884 (15.82%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4268139 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8536278 mates make up the pairs; of these:\n",
      "        8426328 (98.71%) aligned 0 times\n",
      "        62891 (0.74%) aligned exactly 1 time\n",
      "        47059 (0.55%) aligned >1 times\n",
      "33.08% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:08:05] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78541.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78541.bam.\n",
      "[INFO] [Feb 12, 2023 18:08:27] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78541.sam.\n",
      "[INFO] [Feb 12, 2023 18:08:27] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78554.sam\n",
      "[INFO] [Feb 12, 2023 18:08:27] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78554_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78554_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2669923 reads; of these:\n",
      "  2669923 (100.00%) were paired; of these:\n",
      "    2288379 (85.71%) aligned concordantly 0 times\n",
      "    369339 (13.83%) aligned concordantly exactly 1 time\n",
      "    12205 (0.46%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2288379 pairs aligned concordantly 0 times; of these:\n",
      "      125471 (5.48%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2162908 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4325816 mates make up the pairs; of these:\n",
      "        4268316 (98.67%) aligned 0 times\n",
      "        46130 (1.07%) aligned exactly 1 time\n",
      "        11370 (0.26%) aligned >1 times\n",
      "20.07% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:08:57] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78554.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78554.bam.\n",
      "[INFO] [Feb 12, 2023 18:09:03] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78554.sam.\n",
      "[INFO] [Feb 12, 2023 18:09:03] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78627.sam\n",
      "[INFO] [Feb 12, 2023 18:09:03] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78627_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78627_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2817973 reads; of these:\n",
      "  2817973 (100.00%) were paired; of these:\n",
      "    2626324 (93.20%) aligned concordantly 0 times\n",
      "    184232 (6.54%) aligned concordantly exactly 1 time\n",
      "    7417 (0.26%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2626324 pairs aligned concordantly 0 times; of these:\n",
      "      54077 (2.06%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2572247 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5144494 mates make up the pairs; of these:\n",
      "        5112458 (99.38%) aligned 0 times\n",
      "        23369 (0.45%) aligned exactly 1 time\n",
      "        8667 (0.17%) aligned >1 times\n",
      "9.29% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:09:29] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78627.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78627.bam.\n",
      "[INFO] [Feb 12, 2023 18:09:32] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78627.sam.\n",
      "[INFO] [Feb 12, 2023 18:09:33] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78655.sam\n",
      "[INFO] [Feb 12, 2023 18:09:33] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78655_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78655_2.fastq.gz --no-unal -p 12 -k 10\n",
      "5001703 reads; of these:\n",
      "  5001703 (100.00%) were paired; of these:\n",
      "    4515669 (90.28%) aligned concordantly 0 times\n",
      "    469747 (9.39%) aligned concordantly exactly 1 time\n",
      "    16287 (0.33%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4515669 pairs aligned concordantly 0 times; of these:\n",
      "      326741 (7.24%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4188928 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8377856 mates make up the pairs; of these:\n",
      "        8320675 (99.32%) aligned 0 times\n",
      "        25549 (0.30%) aligned exactly 1 time\n",
      "        31632 (0.38%) aligned >1 times\n",
      "16.82% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:10:29] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78655.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78655.bam.\n",
      "[INFO] [Feb 12, 2023 18:10:39] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78655.sam.\n",
      "[INFO] [Feb 12, 2023 18:10:39] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78714.sam\n",
      "[INFO] [Feb 12, 2023 18:10:39] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78714_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78714_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7085374 reads; of these:\n",
      "  7085374 (100.00%) were paired; of these:\n",
      "    6772862 (95.59%) aligned concordantly 0 times\n",
      "    292867 (4.13%) aligned concordantly exactly 1 time\n",
      "    19645 (0.28%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6772862 pairs aligned concordantly 0 times; of these:\n",
      "      70046 (1.03%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6702816 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      13405632 mates make up the pairs; of these:\n",
      "        13350697 (99.59%) aligned 0 times\n",
      "        33902 (0.25%) aligned exactly 1 time\n",
      "        21033 (0.16%) aligned >1 times\n",
      "5.79% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:11:42] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78714.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78714.bam.\n",
      "[INFO] [Feb 12, 2023 18:11:47] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78714.sam.\n",
      "[INFO] [Feb 12, 2023 18:11:47] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78717.sam\n",
      "[INFO] [Feb 12, 2023 18:11:47] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78717_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78717_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3677056 reads; of these:\n",
      "  3677056 (100.00%) were paired; of these:\n",
      "    3230967 (87.87%) aligned concordantly 0 times\n",
      "    433232 (11.78%) aligned concordantly exactly 1 time\n",
      "    12857 (0.35%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3230967 pairs aligned concordantly 0 times; of these:\n",
      "      123586 (3.83%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3107381 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6214762 mates make up the pairs; of these:\n",
      "        6178079 (99.41%) aligned 0 times\n",
      "        24304 (0.39%) aligned exactly 1 time\n",
      "        12379 (0.20%) aligned >1 times\n",
      "15.99% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:12:27] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78717.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78717.bam.\n",
      "[INFO] [Feb 12, 2023 18:12:34] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78717.sam.\n",
      "[INFO] [Feb 12, 2023 18:12:34] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78718.sam\n",
      "[INFO] [Feb 12, 2023 18:12:34] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78718_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78718_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8329356 reads; of these:\n",
      "  8329356 (100.00%) were paired; of these:\n",
      "    8003515 (96.09%) aligned concordantly 0 times\n",
      "    307667 (3.69%) aligned concordantly exactly 1 time\n",
      "    18174 (0.22%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8003515 pairs aligned concordantly 0 times; of these:\n",
      "      106861 (1.34%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7896654 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      15793308 mates make up the pairs; of these:\n",
      "        15736526 (99.64%) aligned 0 times\n",
      "        30661 (0.19%) aligned exactly 1 time\n",
      "        26121 (0.17%) aligned >1 times\n",
      "5.54% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:13:47] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78718.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78718.bam.\n",
      "[INFO] [Feb 12, 2023 18:13:52] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78718.sam.\n",
      "[INFO] [Feb 12, 2023 18:13:52] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78719.sam\n",
      "[INFO] [Feb 12, 2023 18:13:52] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78719_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78719_2.fastq.gz --no-unal -p 12 -k 10\n",
      "4112392 reads; of these:\n",
      "  4112392 (100.00%) were paired; of these:\n",
      "    4075155 (99.09%) aligned concordantly 0 times\n",
      "    28422 (0.69%) aligned concordantly exactly 1 time\n",
      "    8815 (0.21%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4075155 pairs aligned concordantly 0 times; of these:\n",
      "      6461 (0.16%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4068694 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8137388 mates make up the pairs; of these:\n",
      "        8117391 (99.75%) aligned 0 times\n",
      "        10314 (0.13%) aligned exactly 1 time\n",
      "        9683 (0.12%) aligned >1 times\n",
      "1.31% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:14:25] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78719.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78719.bam.\n",
      "[INFO] [Feb 12, 2023 18:14:26] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78719.sam.\n",
      "[INFO] [Feb 12, 2023 18:14:26] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78728.sam\n",
      "[INFO] [Feb 12, 2023 18:14:26] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78728_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78728_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3266950 reads; of these:\n",
      "  3266950 (100.00%) were paired; of these:\n",
      "    3085388 (94.44%) aligned concordantly 0 times\n",
      "    176992 (5.42%) aligned concordantly exactly 1 time\n",
      "    4570 (0.14%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3085388 pairs aligned concordantly 0 times; of these:\n",
      "      437442 (14.18%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2647946 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5295892 mates make up the pairs; of these:\n",
      "        5255156 (99.23%) aligned 0 times\n",
      "        12651 (0.24%) aligned exactly 1 time\n",
      "        28085 (0.53%) aligned >1 times\n",
      "19.57% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:15:02] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78728.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78728.bam.\n",
      "[INFO] [Feb 12, 2023 18:15:10] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78728.sam.\n",
      "[INFO] [Feb 12, 2023 18:15:10] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78734.sam\n",
      "[INFO] [Feb 12, 2023 18:15:10] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78734_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78734_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2383975 reads; of these:\n",
      "  2383975 (100.00%) were paired; of these:\n",
      "    1765172 (74.04%) aligned concordantly 0 times\n",
      "    603686 (25.32%) aligned concordantly exactly 1 time\n",
      "    15117 (0.63%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1765172 pairs aligned concordantly 0 times; of these:\n",
      "      233961 (13.25%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1531211 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3062422 mates make up the pairs; of these:\n",
      "        3010604 (98.31%) aligned 0 times\n",
      "        35318 (1.15%) aligned exactly 1 time\n",
      "        16500 (0.54%) aligned >1 times\n",
      "36.86% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:15:49] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78734.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78734.bam.\n",
      "[INFO] [Feb 12, 2023 18:15:59] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78734.sam.\n",
      "[INFO] [Feb 12, 2023 18:15:59] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78744.sam\n",
      "[INFO] [Feb 12, 2023 18:15:59] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78744_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78744_2.fastq.gz --no-unal -p 12 -k 10\n",
      "4868881 reads; of these:\n",
      "  4868881 (100.00%) were paired; of these:\n",
      "    4567411 (93.81%) aligned concordantly 0 times\n",
      "    290833 (5.97%) aligned concordantly exactly 1 time\n",
      "    10637 (0.22%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4567411 pairs aligned concordantly 0 times; of these:\n",
      "      495360 (10.85%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4072051 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8144102 mates make up the pairs; of these:\n",
      "        8073285 (99.13%) aligned 0 times\n",
      "        25125 (0.31%) aligned exactly 1 time\n",
      "        45692 (0.56%) aligned >1 times\n",
      "17.09% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:16:52] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78744.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78744.bam.\n",
      "[INFO] [Feb 12, 2023 18:17:02] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78744.sam.\n",
      "[INFO] [Feb 12, 2023 18:17:02] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78751.sam\n",
      "[INFO] [Feb 12, 2023 18:17:02] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78751_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78751_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3359116 reads; of these:\n",
      "  3359116 (100.00%) were paired; of these:\n",
      "    3315603 (98.70%) aligned concordantly 0 times\n",
      "    41976 (1.25%) aligned concordantly exactly 1 time\n",
      "    1537 (0.05%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3315603 pairs aligned concordantly 0 times; of these:\n",
      "      427395 (12.89%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2888208 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5776416 mates make up the pairs; of these:\n",
      "        5705110 (98.77%) aligned 0 times\n",
      "        19110 (0.33%) aligned exactly 1 time\n",
      "        52196 (0.90%) aligned >1 times\n",
      "15.08% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:17:35] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78751.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78751.bam.\n",
      "[INFO] [Feb 12, 2023 18:17:40] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78751.sam.\n",
      "[INFO] [Feb 12, 2023 18:17:40] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78773.sam\n",
      "[INFO] [Feb 12, 2023 18:17:40] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78773_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78773_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2011799 reads; of these:\n",
      "  2011799 (100.00%) were paired; of these:\n",
      "    1790629 (89.01%) aligned concordantly 0 times\n",
      "    213946 (10.63%) aligned concordantly exactly 1 time\n",
      "    7224 (0.36%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1790629 pairs aligned concordantly 0 times; of these:\n",
      "      60330 (3.37%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1730299 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3460598 mates make up the pairs; of these:\n",
      "        3439680 (99.40%) aligned 0 times\n",
      "        13720 (0.40%) aligned exactly 1 time\n",
      "        7198 (0.21%) aligned >1 times\n",
      "14.51% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:18:01] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78773.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78773.bam.\n",
      "[INFO] [Feb 12, 2023 18:18:05] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78773.sam.\n",
      "[INFO] [Feb 12, 2023 18:18:05] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78808.sam\n",
      "[INFO] [Feb 12, 2023 18:18:05] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78808_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78808_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3452763 reads; of these:\n",
      "  3452763 (100.00%) were paired; of these:\n",
      "    3340349 (96.74%) aligned concordantly 0 times\n",
      "    107544 (3.11%) aligned concordantly exactly 1 time\n",
      "    4870 (0.14%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3340349 pairs aligned concordantly 0 times; of these:\n",
      "      238786 (7.15%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3101563 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6203126 mates make up the pairs; of these:\n",
      "        6165274 (99.39%) aligned 0 times\n",
      "        11957 (0.19%) aligned exactly 1 time\n",
      "        25895 (0.42%) aligned >1 times\n",
      "10.72% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:18:37] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78808.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78808.bam.\n",
      "[INFO] [Feb 12, 2023 18:18:41] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78808.sam.\n",
      "[INFO] [Feb 12, 2023 18:18:41] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78809.sam\n",
      "[INFO] [Feb 12, 2023 18:18:41] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78809_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78809_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8573025 reads; of these:\n",
      "  8573025 (100.00%) were paired; of these:\n",
      "    8226125 (95.95%) aligned concordantly 0 times\n",
      "    331541 (3.87%) aligned concordantly exactly 1 time\n",
      "    15359 (0.18%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8226125 pairs aligned concordantly 0 times; of these:\n",
      "      575710 (7.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7650415 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      15300830 mates make up the pairs; of these:\n",
      "        15206633 (99.38%) aligned 0 times\n",
      "        30538 (0.20%) aligned exactly 1 time\n",
      "        63659 (0.42%) aligned >1 times\n",
      "11.31% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:20:05] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78809.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78809.bam.\n",
      "[INFO] [Feb 12, 2023 18:20:16] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78809.sam.\n",
      "[INFO] [Feb 12, 2023 18:20:16] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78816.sam\n",
      "[INFO] [Feb 12, 2023 18:20:16] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78816_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78816_2.fastq.gz --no-unal -p 12 -k 10\n",
      "1476020 reads; of these:\n",
      "  1476020 (100.00%) were paired; of these:\n",
      "    1474869 (99.92%) aligned concordantly 0 times\n",
      "    906 (0.06%) aligned concordantly exactly 1 time\n",
      "    245 (0.02%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1474869 pairs aligned concordantly 0 times; of these:\n",
      "      3808 (0.26%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1471061 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      2942122 mates make up the pairs; of these:\n",
      "        2937574 (99.85%) aligned 0 times\n",
      "        749 (0.03%) aligned exactly 1 time\n",
      "        3799 (0.13%) aligned >1 times\n",
      "0.49% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:20:26] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78816.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78816.bam.\n",
      "[INFO] [Feb 12, 2023 18:20:26] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78816.sam.\n",
      "[INFO] [Feb 12, 2023 18:20:26] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78847.sam\n",
      "[INFO] [Feb 12, 2023 18:20:26] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78847_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78847_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2815786 reads; of these:\n",
      "  2815786 (100.00%) were paired; of these:\n",
      "    2721212 (96.64%) aligned concordantly 0 times\n",
      "    88986 (3.16%) aligned concordantly exactly 1 time\n",
      "    5588 (0.20%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2721212 pairs aligned concordantly 0 times; of these:\n",
      "      65177 (2.40%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2656035 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5312070 mates make up the pairs; of these:\n",
      "        5296969 (99.72%) aligned 0 times\n",
      "        2594 (0.05%) aligned exactly 1 time\n",
      "        12507 (0.24%) aligned >1 times\n",
      "5.94% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:20:51] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78847.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78847.bam.\n",
      "[INFO] [Feb 12, 2023 18:20:53] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78847.sam.\n",
      "[INFO] [Feb 12, 2023 18:20:53] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78855.sam\n",
      "[INFO] [Feb 12, 2023 18:20:53] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78855_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78855_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3845201 reads; of these:\n",
      "  3845201 (100.00%) were paired; of these:\n",
      "    3429146 (89.18%) aligned concordantly 0 times\n",
      "    400045 (10.40%) aligned concordantly exactly 1 time\n",
      "    16010 (0.42%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3429146 pairs aligned concordantly 0 times; of these:\n",
      "      109703 (3.20%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3319443 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6638886 mates make up the pairs; of these:\n",
      "        6598457 (99.39%) aligned 0 times\n",
      "        24133 (0.36%) aligned exactly 1 time\n",
      "        16296 (0.25%) aligned >1 times\n",
      "14.20% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:21:34] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78855.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78855.bam.\n",
      "[INFO] [Feb 12, 2023 18:21:41] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78855.sam.\n",
      "[INFO] [Feb 12, 2023 18:21:41] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78856.sam\n",
      "[INFO] [Feb 12, 2023 18:21:41] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78856_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78856_2.fastq.gz --no-unal -p 12 -k 10\n",
      "4904995 reads; of these:\n",
      "  4904995 (100.00%) were paired; of these:\n",
      "    4716054 (96.15%) aligned concordantly 0 times\n",
      "    177981 (3.63%) aligned concordantly exactly 1 time\n",
      "    10960 (0.22%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4716054 pairs aligned concordantly 0 times; of these:\n",
      "      44975 (0.95%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4671079 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      9342158 mates make up the pairs; of these:\n",
      "        9306450 (99.62%) aligned 0 times\n",
      "        24052 (0.26%) aligned exactly 1 time\n",
      "        11656 (0.12%) aligned >1 times\n",
      "5.13% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:22:23] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78856.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78856.bam.\n",
      "[INFO] [Feb 12, 2023 18:22:27] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78856.sam.\n",
      "[INFO] [Feb 12, 2023 18:22:27] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78864.sam\n",
      "[INFO] [Feb 12, 2023 18:22:27] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78864_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78864_2.fastq.gz --no-unal -p 12 -k 10\n",
      "5434731 reads; of these:\n",
      "  5434731 (100.00%) were paired; of these:\n",
      "    5201490 (95.71%) aligned concordantly 0 times\n",
      "    221374 (4.07%) aligned concordantly exactly 1 time\n",
      "    11867 (0.22%) aligned concordantly >1 times\n",
      "    ----\n",
      "    5201490 pairs aligned concordantly 0 times; of these:\n",
      "      120142 (2.31%) aligned discordantly 1 time\n",
      "    ----\n",
      "    5081348 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      10162696 mates make up the pairs; of these:\n",
      "        10124667 (99.63%) aligned 0 times\n",
      "        16521 (0.16%) aligned exactly 1 time\n",
      "        21508 (0.21%) aligned >1 times\n",
      "6.85% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:23:15] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78864.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78864.bam.\n",
      "[INFO] [Feb 12, 2023 18:23:20] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78864.sam.\n",
      "[INFO] [Feb 12, 2023 18:23:20] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78866.sam\n",
      "[INFO] [Feb 12, 2023 18:23:20] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78866_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78866_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6093144 reads; of these:\n",
      "  6093144 (100.00%) were paired; of these:\n",
      "    5369770 (88.13%) aligned concordantly 0 times\n",
      "    688534 (11.30%) aligned concordantly exactly 1 time\n",
      "    34840 (0.57%) aligned concordantly >1 times\n",
      "    ----\n",
      "    5369770 pairs aligned concordantly 0 times; of these:\n",
      "      371974 (6.93%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4997796 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      9995592 mates make up the pairs; of these:\n",
      "        9900961 (99.05%) aligned 0 times\n",
      "        45340 (0.45%) aligned exactly 1 time\n",
      "        49291 (0.49%) aligned >1 times\n",
      "18.75% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:24:34] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78866.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78866.bam.\n",
      "[INFO] [Feb 12, 2023 18:24:47] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78866.sam.\n",
      "[INFO] [Feb 12, 2023 18:24:47] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78872.sam\n",
      "[INFO] [Feb 12, 2023 18:24:47] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78872_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78872_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3283751 reads; of these:\n",
      "  3283751 (100.00%) were paired; of these:\n",
      "    2478349 (75.47%) aligned concordantly 0 times\n",
      "    783102 (23.85%) aligned concordantly exactly 1 time\n",
      "    22300 (0.68%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2478349 pairs aligned concordantly 0 times; of these:\n",
      "      120698 (4.87%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2357651 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4715302 mates make up the pairs; of these:\n",
      "        4663595 (98.90%) aligned 0 times\n",
      "        40992 (0.87%) aligned exactly 1 time\n",
      "        10715 (0.23%) aligned >1 times\n",
      "28.99% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:25:32] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78872.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78872.bam.\n",
      "[INFO] [Feb 12, 2023 18:25:42] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78872.sam.\n",
      "[INFO] [Feb 12, 2023 18:25:42] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78878.sam\n",
      "[INFO] [Feb 12, 2023 18:25:42] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78878_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78878_2.fastq.gz --no-unal -p 12 -k 10\n",
      "4053063 reads; of these:\n",
      "  4053063 (100.00%) were paired; of these:\n",
      "    2723807 (67.20%) aligned concordantly 0 times\n",
      "    1300193 (32.08%) aligned concordantly exactly 1 time\n",
      "    29063 (0.72%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2723807 pairs aligned concordantly 0 times; of these:\n",
      "      292401 (10.74%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2431406 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4862812 mates make up the pairs; of these:\n",
      "        4795317 (98.61%) aligned 0 times\n",
      "        49050 (1.01%) aligned exactly 1 time\n",
      "        18445 (0.38%) aligned >1 times\n",
      "40.84% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:26:50] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78878.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78878.bam.\n",
      "[INFO] [Feb 12, 2023 18:27:07] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78878.sam.\n",
      "[INFO] [Feb 12, 2023 18:27:07] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78898.sam\n",
      "[INFO] [Feb 12, 2023 18:27:07] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78898_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78898_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2930000 reads; of these:\n",
      "  2930000 (100.00%) were paired; of these:\n",
      "    2370469 (80.90%) aligned concordantly 0 times\n",
      "    545990 (18.63%) aligned concordantly exactly 1 time\n",
      "    13541 (0.46%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2370469 pairs aligned concordantly 0 times; of these:\n",
      "      84628 (3.57%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2285841 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4571682 mates make up the pairs; of these:\n",
      "        4538922 (99.28%) aligned 0 times\n",
      "        24739 (0.54%) aligned exactly 1 time\n",
      "        8021 (0.18%) aligned >1 times\n",
      "22.54% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:27:43] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78898.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78898.bam.\n",
      "[INFO] [Feb 12, 2023 18:27:51] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78898.sam.\n",
      "[INFO] [Feb 12, 2023 18:27:51] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78910.sam\n",
      "[INFO] [Feb 12, 2023 18:27:51] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78910_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78910_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7310028 reads; of these:\n",
      "  7310028 (100.00%) were paired; of these:\n",
      "    7071704 (96.74%) aligned concordantly 0 times\n",
      "    222770 (3.05%) aligned concordantly exactly 1 time\n",
      "    15554 (0.21%) aligned concordantly >1 times\n",
      "    ----\n",
      "    7071704 pairs aligned concordantly 0 times; of these:\n",
      "      117801 (1.67%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6953903 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      13907806 mates make up the pairs; of these:\n",
      "        13851005 (99.59%) aligned 0 times\n",
      "        33348 (0.24%) aligned exactly 1 time\n",
      "        23453 (0.17%) aligned >1 times\n",
      "5.26% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:28:53] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78910.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78910.bam.\n",
      "[INFO] [Feb 12, 2023 18:28:58] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78910.sam.\n",
      "[INFO] [Feb 12, 2023 18:28:58] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78923.sam\n",
      "[INFO] [Feb 12, 2023 18:28:58] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78923_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78923_2.fastq.gz --no-unal -p 12 -k 10\n",
      "12141682 reads; of these:\n",
      "  12141682 (100.00%) were paired; of these:\n",
      "    10295118 (84.79%) aligned concordantly 0 times\n",
      "    1796254 (14.79%) aligned concordantly exactly 1 time\n",
      "    50310 (0.41%) aligned concordantly >1 times\n",
      "    ----\n",
      "    10295118 pairs aligned concordantly 0 times; of these:\n",
      "      715428 (6.95%) aligned discordantly 1 time\n",
      "    ----\n",
      "    9579690 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      19159380 mates make up the pairs; of these:\n",
      "        18997338 (99.15%) aligned 0 times\n",
      "        101281 (0.53%) aligned exactly 1 time\n",
      "        60761 (0.32%) aligned >1 times\n",
      "21.77% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:31:32] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78923.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78923.bam.\n",
      "[INFO] [Feb 12, 2023 18:32:05] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78923.sam.\n",
      "[INFO] [Feb 12, 2023 18:32:06] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78928.sam\n",
      "[INFO] [Feb 12, 2023 18:32:06] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78928_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78928_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7411803 reads; of these:\n",
      "  7411803 (100.00%) were paired; of these:\n",
      "    6868011 (92.66%) aligned concordantly 0 times\n",
      "    528517 (7.13%) aligned concordantly exactly 1 time\n",
      "    15275 (0.21%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6868011 pairs aligned concordantly 0 times; of these:\n",
      "      176436 (2.57%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6691575 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      13383150 mates make up the pairs; of these:\n",
      "        13265026 (99.12%) aligned 0 times\n",
      "        97698 (0.73%) aligned exactly 1 time\n",
      "        20426 (0.15%) aligned >1 times\n",
      "10.51% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:33:19] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78928.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78928.bam.\n",
      "[INFO] [Feb 12, 2023 18:33:27] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78928.sam.\n",
      "[INFO] [Feb 12, 2023 18:33:27] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78944.sam\n",
      "[INFO] [Feb 12, 2023 18:33:27] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78944_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78944_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2443284 reads; of these:\n",
      "  2443284 (100.00%) were paired; of these:\n",
      "    1741422 (71.27%) aligned concordantly 0 times\n",
      "    686194 (28.08%) aligned concordantly exactly 1 time\n",
      "    15668 (0.64%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1741422 pairs aligned concordantly 0 times; of these:\n",
      "      98250 (5.64%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1643172 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3286344 mates make up the pairs; of these:\n",
      "        3245738 (98.76%) aligned 0 times\n",
      "        32523 (0.99%) aligned exactly 1 time\n",
      "        8083 (0.25%) aligned >1 times\n",
      "33.58% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:34:04] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78944.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78944.bam.\n",
      "[INFO] [Feb 12, 2023 18:34:14] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78944.sam.\n",
      "[INFO] [Feb 12, 2023 18:34:14] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78961.sam\n",
      "[INFO] [Feb 12, 2023 18:34:14] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78961_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78961_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2015397 reads; of these:\n",
      "  2015397 (100.00%) were paired; of these:\n",
      "    1840783 (91.34%) aligned concordantly 0 times\n",
      "    167156 (8.29%) aligned concordantly exactly 1 time\n",
      "    7458 (0.37%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1840783 pairs aligned concordantly 0 times; of these:\n",
      "      28594 (1.55%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1812189 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3624378 mates make up the pairs; of these:\n",
      "        3606595 (99.51%) aligned 0 times\n",
      "        12888 (0.36%) aligned exactly 1 time\n",
      "        4895 (0.14%) aligned >1 times\n",
      "10.52% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:34:34] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78961.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78961.bam.\n",
      "[INFO] [Feb 12, 2023 18:34:36] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78961.sam.\n",
      "[INFO] [Feb 12, 2023 18:34:36] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78962.sam\n",
      "[INFO] [Feb 12, 2023 18:34:36] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78962_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78962_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2477218 reads; of these:\n",
      "  2477218 (100.00%) were paired; of these:\n",
      "    2370338 (95.69%) aligned concordantly 0 times\n",
      "    100432 (4.05%) aligned concordantly exactly 1 time\n",
      "    6448 (0.26%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2370338 pairs aligned concordantly 0 times; of these:\n",
      "      17933 (0.76%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2352405 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4704810 mates make up the pairs; of these:\n",
      "        4679830 (99.47%) aligned 0 times\n",
      "        19042 (0.40%) aligned exactly 1 time\n",
      "        5938 (0.13%) aligned >1 times\n",
      "5.54% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:34:58] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78962.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78962.bam.\n",
      "[INFO] [Feb 12, 2023 18:35:00] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78962.sam.\n",
      "[INFO] [Feb 12, 2023 18:35:00] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78973.sam\n",
      "[INFO] [Feb 12, 2023 18:35:00] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78973_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78973_2.fastq.gz --no-unal -p 12 -k 10\n",
      "1602410 reads; of these:\n",
      "  1602410 (100.00%) were paired; of these:\n",
      "    1372382 (85.64%) aligned concordantly 0 times\n",
      "    222094 (13.86%) aligned concordantly exactly 1 time\n",
      "    7934 (0.50%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1372382 pairs aligned concordantly 0 times; of these:\n",
      "      27587 (2.01%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1344795 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      2689590 mates make up the pairs; of these:\n",
      "        2675090 (99.46%) aligned 0 times\n",
      "        9826 (0.37%) aligned exactly 1 time\n",
      "        4674 (0.17%) aligned >1 times\n",
      "16.53% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:35:18] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78973.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78973.bam.\n",
      "[INFO] [Feb 12, 2023 18:35:20] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78973.sam.\n",
      "[INFO] [Feb 12, 2023 18:35:21] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78975.sam\n",
      "[INFO] [Feb 12, 2023 18:35:21] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78975_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78975_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3507503 reads; of these:\n",
      "  3507503 (100.00%) were paired; of these:\n",
      "    2419193 (68.97%) aligned concordantly 0 times\n",
      "    1073416 (30.60%) aligned concordantly exactly 1 time\n",
      "    14894 (0.42%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2419193 pairs aligned concordantly 0 times; of these:\n",
      "      164679 (6.81%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2254514 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4509028 mates make up the pairs; of these:\n",
      "        4377190 (97.08%) aligned 0 times\n",
      "        124826 (2.77%) aligned exactly 1 time\n",
      "        7012 (0.16%) aligned >1 times\n",
      "37.60% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:36:19] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78975.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78975.bam.\n",
      "[INFO] [Feb 12, 2023 18:36:32] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78975.sam.\n",
      "[INFO] [Feb 12, 2023 18:36:32] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78977.sam\n",
      "[INFO] [Feb 12, 2023 18:36:32] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78977_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78977_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2010108 reads; of these:\n",
      "  2010108 (100.00%) were paired; of these:\n",
      "    1888041 (93.93%) aligned concordantly 0 times\n",
      "    114771 (5.71%) aligned concordantly exactly 1 time\n",
      "    7296 (0.36%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1888041 pairs aligned concordantly 0 times; of these:\n",
      "      14857 (0.79%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1873184 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3746368 mates make up the pairs; of these:\n",
      "        3725613 (99.45%) aligned 0 times\n",
      "        15459 (0.41%) aligned exactly 1 time\n",
      "        5296 (0.14%) aligned >1 times\n",
      "7.33% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:36:50] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78977.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78977.bam.\n",
      "[INFO] [Feb 12, 2023 18:36:53] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78977.sam.\n",
      "[INFO] [Feb 12, 2023 18:36:53] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78983.sam\n",
      "[INFO] [Feb 12, 2023 18:36:53] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78983_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G78983_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2980138 reads; of these:\n",
      "  2980138 (100.00%) were paired; of these:\n",
      "    2815068 (94.46%) aligned concordantly 0 times\n",
      "    156708 (5.26%) aligned concordantly exactly 1 time\n",
      "    8362 (0.28%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2815068 pairs aligned concordantly 0 times; of these:\n",
      "      30290 (1.08%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2784778 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5569556 mates make up the pairs; of these:\n",
      "        5539332 (99.46%) aligned 0 times\n",
      "        22972 (0.41%) aligned exactly 1 time\n",
      "        7252 (0.13%) aligned >1 times\n",
      "7.06% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:37:20] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78983.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78983.bam.\n",
      "[INFO] [Feb 12, 2023 18:37:23] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78983.sam.\n",
      "[INFO] [Feb 12, 2023 18:37:23] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79001.sam\n",
      "[INFO] [Feb 12, 2023 18:37:23] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79001_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79001_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2929737 reads; of these:\n",
      "  2929737 (100.00%) were paired; of these:\n",
      "    2037306 (69.54%) aligned concordantly 0 times\n",
      "    872883 (29.79%) aligned concordantly exactly 1 time\n",
      "    19548 (0.67%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2037306 pairs aligned concordantly 0 times; of these:\n",
      "      169925 (8.34%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1867381 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3734762 mates make up the pairs; of these:\n",
      "        3686336 (98.70%) aligned 0 times\n",
      "        37175 (1.00%) aligned exactly 1 time\n",
      "        11251 (0.30%) aligned >1 times\n",
      "37.09% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:38:11] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79001.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79001.bam.\n",
      "[INFO] [Feb 12, 2023 18:38:24] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79001.sam.\n",
      "[INFO] [Feb 12, 2023 18:38:24] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79007.sam\n",
      "[INFO] [Feb 12, 2023 18:38:24] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79007_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79007_2.fastq.gz --no-unal -p 12 -k 10\n",
      "1936403 reads; of these:\n",
      "  1936403 (100.00%) were paired; of these:\n",
      "    1704054 (88.00%) aligned concordantly 0 times\n",
      "    225439 (11.64%) aligned concordantly exactly 1 time\n",
      "    6910 (0.36%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1704054 pairs aligned concordantly 0 times; of these:\n",
      "      27784 (1.63%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1676270 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3352540 mates make up the pairs; of these:\n",
      "        3309292 (98.71%) aligned 0 times\n",
      "        38185 (1.14%) aligned exactly 1 time\n",
      "        5063 (0.15%) aligned >1 times\n",
      "14.55% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:38:45] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79007.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79007.bam.\n",
      "[INFO] [Feb 12, 2023 18:38:48] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79007.sam.\n",
      "[INFO] [Feb 12, 2023 18:38:48] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79017.sam\n",
      "[INFO] [Feb 12, 2023 18:38:48] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79017_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79017_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2286325 reads; of these:\n",
      "  2286325 (100.00%) were paired; of these:\n",
      "    2147792 (93.94%) aligned concordantly 0 times\n",
      "    132860 (5.81%) aligned concordantly exactly 1 time\n",
      "    5673 (0.25%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2147792 pairs aligned concordantly 0 times; of these:\n",
      "      26602 (1.24%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2121190 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      4242380 mates make up the pairs; of these:\n",
      "        4215337 (99.36%) aligned 0 times\n",
      "        21468 (0.51%) aligned exactly 1 time\n",
      "        5575 (0.13%) aligned >1 times\n",
      "7.81% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:39:09] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79017.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79017.bam.\n",
      "[INFO] [Feb 12, 2023 18:39:12] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79017.sam.\n",
      "[INFO] [Feb 12, 2023 18:39:12] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79018.sam\n",
      "[INFO] [Feb 12, 2023 18:39:12] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79018_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79018_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3136218 reads; of these:\n",
      "  3136218 (100.00%) were paired; of these:\n",
      "    2133094 (68.01%) aligned concordantly 0 times\n",
      "    987558 (31.49%) aligned concordantly exactly 1 time\n",
      "    15566 (0.50%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2133094 pairs aligned concordantly 0 times; of these:\n",
      "      169152 (7.93%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1963942 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3927884 mates make up the pairs; of these:\n",
      "        3897085 (99.22%) aligned 0 times\n",
      "        21353 (0.54%) aligned exactly 1 time\n",
      "        9446 (0.24%) aligned >1 times\n",
      "37.87% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:40:02] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79018.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79018.bam.\n",
      "[INFO] [Feb 12, 2023 18:40:17] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79018.sam.\n",
      "[INFO] [Feb 12, 2023 18:40:17] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79021.sam\n",
      "[INFO] [Feb 12, 2023 18:40:17] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79021_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79021_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3986484 reads; of these:\n",
      "  3986484 (100.00%) were paired; of these:\n",
      "    3551037 (89.08%) aligned concordantly 0 times\n",
      "    422034 (10.59%) aligned concordantly exactly 1 time\n",
      "    13413 (0.34%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3551037 pairs aligned concordantly 0 times; of these:\n",
      "      70809 (1.99%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3480228 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6960456 mates make up the pairs; of these:\n",
      "        6927448 (99.53%) aligned 0 times\n",
      "        23794 (0.34%) aligned exactly 1 time\n",
      "        9214 (0.13%) aligned >1 times\n",
      "13.11% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:40:58] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79021.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79021.bam.\n",
      "[INFO] [Feb 12, 2023 18:41:05] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79021.sam.\n",
      "[INFO] [Feb 12, 2023 18:41:05] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79056.sam\n",
      "[INFO] [Feb 12, 2023 18:41:05] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79056_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79056_2.fastq.gz --no-unal -p 12 -k 10\n",
      "3096691 reads; of these:\n",
      "  3096691 (100.00%) were paired; of these:\n",
      "    2642435 (85.33%) aligned concordantly 0 times\n",
      "    440521 (14.23%) aligned concordantly exactly 1 time\n",
      "    13735 (0.44%) aligned concordantly >1 times\n",
      "    ----\n",
      "    2642435 pairs aligned concordantly 0 times; of these:\n",
      "      74668 (2.83%) aligned discordantly 1 time\n",
      "    ----\n",
      "    2567767 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      5135534 mates make up the pairs; of these:\n",
      "        5100231 (99.31%) aligned 0 times\n",
      "        25486 (0.50%) aligned exactly 1 time\n",
      "        9817 (0.19%) aligned >1 times\n",
      "17.65% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:41:42] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79056.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79056.bam.\n",
      "[INFO] [Feb 12, 2023 18:41:49] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79056.sam.\n",
      "[INFO] [Feb 12, 2023 18:41:49] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79066.sam\n",
      "[INFO] [Feb 12, 2023 18:41:49] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79066_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G79066_2.fastq.gz --no-unal -p 12 -k 10\n",
      "986361 reads; of these:\n",
      "  986361 (100.00%) were paired; of these:\n",
      "    818980 (83.03%) aligned concordantly 0 times\n",
      "    163647 (16.59%) aligned concordantly exactly 1 time\n",
      "    3734 (0.38%) aligned concordantly >1 times\n",
      "    ----\n",
      "    818980 pairs aligned concordantly 0 times; of these:\n",
      "      105677 (12.90%) aligned discordantly 1 time\n",
      "    ----\n",
      "    713303 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      1426606 mates make up the pairs; of these:\n",
      "        1403786 (98.40%) aligned 0 times\n",
      "        16527 (1.16%) aligned exactly 1 time\n",
      "        6293 (0.44%) aligned >1 times\n",
      "28.84% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:42:02] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79066.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79066.bam.\n",
      "[INFO] [Feb 12, 2023 18:42:06] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G79066.sam.\n",
      "[INFO] [Feb 12, 2023 18:42:06] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80095.sam\n",
      "[INFO] [Feb 12, 2023 18:42:06] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80095_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80095_2.fastq.gz --no-unal -p 12 -k 10\n",
      "15231672 reads; of these:\n",
      "  15231672 (100.00%) were paired; of these:\n",
      "    12346459 (81.06%) aligned concordantly 0 times\n",
      "    2837325 (18.63%) aligned concordantly exactly 1 time\n",
      "    47888 (0.31%) aligned concordantly >1 times\n",
      "    ----\n",
      "    12346459 pairs aligned concordantly 0 times; of these:\n",
      "      1061488 (8.60%) aligned discordantly 1 time\n",
      "    ----\n",
      "    11284971 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      22569942 mates make up the pairs; of these:\n",
      "        22176579 (98.26%) aligned 0 times\n",
      "        333515 (1.48%) aligned exactly 1 time\n",
      "        59848 (0.27%) aligned >1 times\n",
      "27.20% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:45:44] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80095.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80095.bam.\n",
      "[INFO] [Feb 12, 2023 18:46:37] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80095.sam.\n",
      "[INFO] [Feb 12, 2023 18:46:37] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80111.sam\n",
      "[INFO] [Feb 12, 2023 18:46:37] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80111_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80111_2.fastq.gz --no-unal -p 12 -k 10\n",
      "23576909 reads; of these:\n",
      "  23576909 (100.00%) were paired; of these:\n",
      "    17428065 (73.92%) aligned concordantly 0 times\n",
      "    6072280 (25.76%) aligned concordantly exactly 1 time\n",
      "    76564 (0.32%) aligned concordantly >1 times\n",
      "    ----\n",
      "    17428065 pairs aligned concordantly 0 times; of these:\n",
      "      2019646 (11.59%) aligned discordantly 1 time\n",
      "    ----\n",
      "    15408419 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      30816838 mates make up the pairs; of these:\n",
      "        30097908 (97.67%) aligned 0 times\n",
      "        642289 (2.08%) aligned exactly 1 time\n",
      "        76641 (0.25%) aligned >1 times\n",
      "36.17% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:53:05] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80111.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80111.bam.\n",
      "[INFO] [Feb 12, 2023 18:55:03] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80111.sam.\n",
      "[INFO] [Feb 12, 2023 18:55:03] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80126.sam\n",
      "[INFO] [Feb 12, 2023 18:55:03] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80126_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80126_2.fastq.gz --no-unal -p 12 -k 10\n",
      "1013949 reads; of these:\n",
      "  1013949 (100.00%) were paired; of these:\n",
      "    854342 (84.26%) aligned concordantly 0 times\n",
      "    154408 (15.23%) aligned concordantly exactly 1 time\n",
      "    5199 (0.51%) aligned concordantly >1 times\n",
      "    ----\n",
      "    854342 pairs aligned concordantly 0 times; of these:\n",
      "      130880 (15.32%) aligned discordantly 1 time\n",
      "    ----\n",
      "    723462 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      1446924 mates make up the pairs; of these:\n",
      "        1420431 (98.17%) aligned 0 times\n",
      "        16644 (1.15%) aligned exactly 1 time\n",
      "        9849 (0.68%) aligned >1 times\n",
      "29.96% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:55:18] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80126.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80126.bam.\n",
      "[INFO] [Feb 12, 2023 18:55:22] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80126.sam.\n",
      "[INFO] [Feb 12, 2023 18:55:22] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80157.sam\n",
      "[INFO] [Feb 12, 2023 18:55:22] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80157_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80157_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7157844 reads; of these:\n",
      "  7157844 (100.00%) were paired; of these:\n",
      "    6706527 (93.69%) aligned concordantly 0 times\n",
      "    430529 (6.01%) aligned concordantly exactly 1 time\n",
      "    20788 (0.29%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6706527 pairs aligned concordantly 0 times; of these:\n",
      "      151240 (2.26%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6555287 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      13110574 mates make up the pairs; of these:\n",
      "        13028312 (99.37%) aligned 0 times\n",
      "        57458 (0.44%) aligned exactly 1 time\n",
      "        24804 (0.19%) aligned >1 times\n",
      "8.99% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:56:30] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80157.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80157.bam.\n",
      "[INFO] [Feb 12, 2023 18:56:38] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80157.sam.\n",
      "[INFO] [Feb 12, 2023 18:56:38] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80388.sam\n",
      "[INFO] [Feb 12, 2023 18:56:38] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80388_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80388_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6877833 reads; of these:\n",
      "  6877833 (100.00%) were paired; of these:\n",
      "    5754598 (83.67%) aligned concordantly 0 times\n",
      "    1095780 (15.93%) aligned concordantly exactly 1 time\n",
      "    27455 (0.40%) aligned concordantly >1 times\n",
      "    ----\n",
      "    5754598 pairs aligned concordantly 0 times; of these:\n",
      "      809803 (14.07%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4944795 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      9889590 mates make up the pairs; of these:\n",
      "        9774729 (98.84%) aligned 0 times\n",
      "        65442 (0.66%) aligned exactly 1 time\n",
      "        49419 (0.50%) aligned >1 times\n",
      "28.94% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 18:58:56] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80388.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80388.bam.\n",
      "[INFO] [Feb 12, 2023 18:59:22] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80388.sam.\n",
      "[INFO] [Feb 12, 2023 18:59:22] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80392.sam\n",
      "[INFO] [Feb 12, 2023 18:59:22] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80392_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80392_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6026521 reads; of these:\n",
      "  6026521 (100.00%) were paired; of these:\n",
      "    4821085 (80.00%) aligned concordantly 0 times\n",
      "    1177436 (19.54%) aligned concordantly exactly 1 time\n",
      "    28000 (0.46%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4821085 pairs aligned concordantly 0 times; of these:\n",
      "      830329 (17.22%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3990756 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      7981512 mates make up the pairs; of these:\n",
      "        7884481 (98.78%) aligned 0 times\n",
      "        50256 (0.63%) aligned exactly 1 time\n",
      "        46775 (0.59%) aligned >1 times\n",
      "34.59% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:00:58] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80392.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80392.bam.\n",
      "[INFO] [Feb 12, 2023 19:01:24] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80392.sam.\n",
      "[INFO] [Feb 12, 2023 19:01:24] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80439.sam\n",
      "[INFO] [Feb 12, 2023 19:01:24] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80439_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80439_2.fastq.gz --no-unal -p 12 -k 10\n",
      "9858854 reads; of these:\n",
      "  9858854 (100.00%) were paired; of these:\n",
      "    8520121 (86.42%) aligned concordantly 0 times\n",
      "    1306499 (13.25%) aligned concordantly exactly 1 time\n",
      "    32234 (0.33%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8520121 pairs aligned concordantly 0 times; of these:\n",
      "      888045 (10.42%) aligned discordantly 1 time\n",
      "    ----\n",
      "    7632076 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      15264152 mates make up the pairs; of these:\n",
      "        15040193 (98.53%) aligned 0 times\n",
      "        159037 (1.04%) aligned exactly 1 time\n",
      "        64922 (0.43%) aligned >1 times\n",
      "23.72% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:03:35] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80439.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80439.bam.\n",
      "[INFO] [Feb 12, 2023 19:04:02] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80439.sam.\n",
      "[INFO] [Feb 12, 2023 19:04:02] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80448.sam\n",
      "[INFO] [Feb 12, 2023 19:04:02] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80448_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80448_2.fastq.gz --no-unal -p 12 -k 10\n",
      "10037129 reads; of these:\n",
      "  10037129 (100.00%) were paired; of these:\n",
      "    8753387 (87.21%) aligned concordantly 0 times\n",
      "    1246968 (12.42%) aligned concordantly exactly 1 time\n",
      "    36774 (0.37%) aligned concordantly >1 times\n",
      "    ----\n",
      "    8753387 pairs aligned concordantly 0 times; of these:\n",
      "      526162 (6.01%) aligned discordantly 1 time\n",
      "    ----\n",
      "    8227225 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      16454450 mates make up the pairs; of these:\n",
      "        16357674 (99.41%) aligned 0 times\n",
      "        49866 (0.30%) aligned exactly 1 time\n",
      "        46910 (0.29%) aligned >1 times\n",
      "18.51% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:06:03] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80448.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80448.bam.\n",
      "[INFO] [Feb 12, 2023 19:06:21] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80448.sam.\n",
      "[INFO] [Feb 12, 2023 19:06:22] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80450.sam\n",
      "[INFO] [Feb 12, 2023 19:06:22] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80450_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80450_2.fastq.gz --no-unal -p 12 -k 10\n",
      "8388462 reads; of these:\n",
      "  8388462 (100.00%) were paired; of these:\n",
      "    6804925 (81.12%) aligned concordantly 0 times\n",
      "    1539280 (18.35%) aligned concordantly exactly 1 time\n",
      "    44257 (0.53%) aligned concordantly >1 times\n",
      "    ----\n",
      "    6804925 pairs aligned concordantly 0 times; of these:\n",
      "      734077 (10.79%) aligned discordantly 1 time\n",
      "    ----\n",
      "    6070848 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      12141696 mates make up the pairs; of these:\n",
      "        12011527 (98.93%) aligned 0 times\n",
      "        73601 (0.61%) aligned exactly 1 time\n",
      "        56568 (0.47%) aligned >1 times\n",
      "28.40% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:08:27] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80450.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80450.bam.\n",
      "[INFO] [Feb 12, 2023 19:08:53] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80450.sam.\n",
      "[INFO] [Feb 12, 2023 19:08:53] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80530.sam\n",
      "[INFO] [Feb 12, 2023 19:08:53] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80530_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80530_2.fastq.gz --no-unal -p 12 -k 10\n",
      "7800027 reads; of these:\n",
      "  7800027 (100.00%) were paired; of these:\n",
      "    5172240 (66.31%) aligned concordantly 0 times\n",
      "    2578405 (33.06%) aligned concordantly exactly 1 time\n",
      "    49382 (0.63%) aligned concordantly >1 times\n",
      "    ----\n",
      "    5172240 pairs aligned concordantly 0 times; of these:\n",
      "      1052511 (20.35%) aligned discordantly 1 time\n",
      "    ----\n",
      "    4119729 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      8239458 mates make up the pairs; of these:\n",
      "        8073429 (97.98%) aligned 0 times\n",
      "        115165 (1.40%) aligned exactly 1 time\n",
      "        50864 (0.62%) aligned >1 times\n",
      "48.25% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:11:20] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80530.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80530.bam.\n",
      "[INFO] [Feb 12, 2023 19:12:03] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80530.sam.\n",
      "[INFO] [Feb 12, 2023 19:12:03] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80535.sam\n",
      "[INFO] [Feb 12, 2023 19:12:03] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80535_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80535_2.fastq.gz --no-unal -p 12 -k 10\n",
      "6199572 reads; of these:\n",
      "  6199572 (100.00%) were paired; of these:\n",
      "    4525960 (73.00%) aligned concordantly 0 times\n",
      "    1637705 (26.42%) aligned concordantly exactly 1 time\n",
      "    35907 (0.58%) aligned concordantly >1 times\n",
      "    ----\n",
      "    4525960 pairs aligned concordantly 0 times; of these:\n",
      "      653513 (14.44%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3872447 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      7744894 mates make up the pairs; of these:\n",
      "        7641800 (98.67%) aligned 0 times\n",
      "        67445 (0.87%) aligned exactly 1 time\n",
      "        35649 (0.46%) aligned >1 times\n",
      "38.37% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:13:47] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80535.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80535.bam.\n",
      "[INFO] [Feb 12, 2023 19:14:12] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80535.sam.\n",
      "[INFO] [Feb 12, 2023 19:14:12] [coptr.read_mapper] Mapping /home/phil/DATA/diabimmune/wgs_aptr/reads to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80621.sam\n",
      "[INFO] [Feb 12, 2023 19:14:12] [coptr.read_mapper] bowtie2 -x /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/index/seqs -1 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80621_1.fastq.gz -2 /home/phil/DATA/diabimmune/wgs_aptr/reads/G80621_2.fastq.gz --no-unal -p 12 -k 10\n",
      "2380058 reads; of these:\n",
      "  2380058 (100.00%) were paired; of these:\n",
      "    1923303 (80.81%) aligned concordantly 0 times\n",
      "    443761 (18.64%) aligned concordantly exactly 1 time\n",
      "    12994 (0.55%) aligned concordantly >1 times\n",
      "    ----\n",
      "    1923303 pairs aligned concordantly 0 times; of these:\n",
      "      105448 (5.48%) aligned discordantly 1 time\n",
      "    ----\n",
      "    1817855 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      3635710 mates make up the pairs; of these:\n",
      "        3562167 (97.98%) aligned 0 times\n",
      "        64721 (1.78%) aligned exactly 1 time\n",
      "        8822 (0.24%) aligned >1 times\n",
      "25.17% overall alignment rate\n",
      "[INFO] [Feb 12, 2023 19:14:43] [coptr.read_mapper] Converting /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80621.sam to /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80621.bam.\n",
      "[INFO] [Feb 12, 2023 19:14:50] [coptr.read_mapper] Cleaning up /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G80621.sam.\n",
      "[INFO] [Feb 12, 2023 19:15:03] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45050.bam.\n",
      "[INFO] [Feb 12, 2023 19:15:09] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:15:14] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:15:23] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:15:27] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:15:28] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45054.bam.\n",
      "[INFO] [Feb 12, 2023 19:15:35] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:15:44] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:16:00] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:16:09] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:16:09] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45056.bam.\n",
      "[INFO] [Feb 12, 2023 19:16:24] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:16:41] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:17:13] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:17:30] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:17:30] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45067.bam.\n",
      "[INFO] [Feb 12, 2023 19:17:34] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:17:40] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:17:49] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:17:54] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:17:54] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45072.bam.\n",
      "[INFO] [Feb 12, 2023 19:18:09] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:18:28] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:19:02] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:19:20] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:19:20] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45087.bam.\n",
      "[INFO] [Feb 12, 2023 19:19:36] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:19:56] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:20:32] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:20:51] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:20:51] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45092.bam.\n",
      "[INFO] [Feb 12, 2023 19:21:00] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:21:11] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:21:40] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:21:41] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45100.bam.\n",
      "[INFO] [Feb 12, 2023 19:21:57] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:22:17] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:22:54] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:23:13] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:23:13] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45106.bam.\n",
      "[INFO] [Feb 12, 2023 19:23:20] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:23:30] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:23:46] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:23:55] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:23:56] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G45272.bam.\n",
      "[INFO] [Feb 12, 2023 19:24:09] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:24:23] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:24:49] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:25:02] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:25:02] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69166.bam.\n",
      "[INFO] [Feb 12, 2023 19:25:18] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:25:38] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:26:15] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:26:34] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:26:34] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69167.bam.\n",
      "[INFO] [Feb 12, 2023 19:26:41] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:26:52] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:27:10] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:27:20] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:27:20] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69178.bam.\n",
      "[INFO] [Feb 12, 2023 19:27:33] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:27:49] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:28:19] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:28:35] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:28:35] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69187.bam.\n",
      "[INFO] [Feb 12, 2023 19:28:50] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:29:10] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:29:49] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:30:09] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:30:09] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69192.bam.\n",
      "[INFO] [Feb 12, 2023 19:30:27] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:30:53] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:31:39] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:32:03] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:32:03] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69193.bam.\n",
      "[INFO] [Feb 12, 2023 19:32:26] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:32:55] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:33:50] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:34:18] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:34:18] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69206.bam.\n",
      "[INFO] [Feb 12, 2023 19:34:37] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:35:02] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:35:49] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:36:14] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:36:15] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69211.bam.\n",
      "[INFO] [Feb 12, 2023 19:36:29] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:36:48] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:37:24] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:37:43] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:37:43] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69212.bam.\n",
      "[INFO] [Feb 12, 2023 19:37:53] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:38:07] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:38:32] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:38:45] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:38:45] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G69227.bam.\n",
      "[INFO] [Feb 12, 2023 19:39:18] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:40:02] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:41:23] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:42:05] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:42:06] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78487.bam.\n",
      "[INFO] [Feb 12, 2023 19:42:10] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:42:16] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:42:26] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:42:32] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:42:32] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78494.bam.\n",
      "[INFO] [Feb 12, 2023 19:42:34] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:42:37] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:42:43] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:42:46] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:42:47] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78509.bam.\n",
      "[INFO] [Feb 12, 2023 19:42:55] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:43:06] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:43:25] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:43:34] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:43:34] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78517.bam.\n",
      "[INFO] [Feb 12, 2023 19:43:45] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:43:58] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:44:22] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:44:34] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:44:34] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78528.bam.\n",
      "[INFO] [Feb 12, 2023 19:44:43] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:44:54] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:45:14] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:45:24] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:45:24] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78541.bam.\n",
      "[INFO] [Feb 12, 2023 19:45:38] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:45:57] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:46:30] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:46:48] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:46:48] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78554.bam.\n",
      "[INFO] [Feb 12, 2023 19:46:52] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:46:57] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:47:11] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:47:12] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78627.bam.\n",
      "[INFO] [Feb 12, 2023 19:47:13] [coptr.bam_processor] Determining reference genomes.\n",
      "[INFO] [Feb 12, 2023 19:47:16] [coptr.bam_processor] Collecting multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:47:20] [coptr.bam_processor] Assigning multi-mapped reads.\n",
      "[INFO] [Feb 12, 2023 19:47:23] [coptr.bam_processor] Grouping reads by reference genome.\n",
      "[INFO] [Feb 12, 2023 19:47:23] [coptr.bam_processor] Processing /home/phil/DATA/diabimmune/wgs_aptr/coptr_out/bam/G78655.bam.\n"
     ]
    }
   ],
   "source": [
    "# Run coPTR\n",
    "import os\n",
    "\n",
    "outdir = \"/home/phil/DATA/diabimmune/wgs_aptr/coptr_out\"\n",
    "readdir = \"/home/phil/DATA/diabimmune/wgs_aptr/reads\"\n",
    "\n",
    "# Make dirs\n",
    "! mkdir -p {outdir}/bam\n",
    "! mkdir -p {outdir}/coverage_maps\n",
    "\n",
    "# coPTR assumes reads end in _1 and _2, not _R1 and _R2 (idempotent batch rename):\n",
    "for file in os.listdir(readdir):\n",
    "    if file.endswith(\"_R1.fastq.gz\"):\n",
    "        ! mv {readdir}/{file} {readdir}/{file.replace(\"_R1\", \"_1\")}\n",
    "    if file.endswith(\"_R2.fastq.gz\"):\n",
    "        ! mv {readdir}/{file} {readdir}/{file.replace(\"_R2\", \"_2\")}\n",
    "\n",
    "# coPTR commands\n",
    "! coptr map {outdir}/index/seqs {readdir} {outdir}/bam --threads 12 --paired\n",
    "! coptr extract {outdir}/bam {outdir}/coverage_maps\n",
    "! coptr estimate {outdir}/coverage_maps {outdir}/coptr_log2_ptrs.tsv\n",
    "! coptr count {outdir}/coverage_maps {outdir}/coptr_counts.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "coptr",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b057563f3ba3f990268c3b86bdb7e377eb054a139b9a5bb4d57860e24c8e31bc"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
